File: expense-tracker/app/src/main/java/com/farhan/paisatrack/data/Repository.kt
package com.farhan.paisatrack.data

import kotlinx.coroutines.flow.Flow

data class AccountBalance(val account: Account, val balance: Double)

data class PeopleSummary(
    val contactName: String,
    val contactKey: String?,
    /** positive = they owe you, negative = you owe them */
    val net: Double,
    val toReceive: Double,
    val toPay: Double,
    val dueAfterSalary: Boolean,
    val debts: List<Debt>
)

class Repository(private val db: AppDatabase) {

    val accountDao = db.accountDao()
    val categoryDao = db.categoryDao()
    val txnDao = db.txnDao()
    val debtDao = db.debtDao()
    val investmentDao = db.investmentDao()

    fun accounts(): Flow<List<Account>> = accountDao.active()
    fun allAccounts(): Flow<List<Account>> = accountDao.all()
    fun categories(): Flow<List<Category>> = categoryDao.all()
    fun categoriesByType(type: TxnType): Flow<List<Category>> = categoryDao.byType(type)
    fun confirmedTxns(): Flow<List<Txn>> = txnDao.confirmed()
    fun pendingTxns(): Flow<List<Txn>> = txnDao.pending()
    fun debts(): Flow<List<Debt>> = debtDao.all()

    suspend fun accountById(id: Long) = accountDao.byId(id)
    suspend fun txnById(id: Long) = txnDao.byId(id)
    suspend fun debtById(id: Long) = debtDao.byId(id)

    suspend fun upsertAccount(a: Account) = accountDao.insert(a)
    suspend fun updateAccount(a: Account) = accountDao.update(a)
    suspend fun deleteAccount(a: Account) = accountDao.delete(a)

    suspend fun upsertCategory(c: Category) = categoryDao.insert(c)
    suspend fun updateCategory(c: Category) = categoryDao.update(c)
    suspend fun deleteCategory(c: Category) = categoryDao.delete(c)

    suspend fun upsertTxn(t: Txn) = txnDao.insert(t)
    suspend fun updateTxn(t: Txn) = txnDao.update(t)
    suspend fun deleteTxn(t: Txn) = txnDao.delete(t)
    suspend fun deleteTxnById(id: Long) = txnDao.deleteById(id)
    suspend fun refExists(ref: String) = txnDao.countByRef(ref) > 0

    suspend fun upsertDebt(d: Debt) = debtDao.insert(d)
    suspend fun updateDebt(d: Debt) = debtDao.update(d)
    suspend fun deleteDebt(d: Debt) = debtDao.delete(d)

    fun investments(): Flow<List<Investment>> = investmentDao.all()
    suspend fun investmentById(id: Long) = investmentDao.byId(id)
    suspend fun upsertInvestment(i: Investment) = investmentDao.insert(i)
    suspend fun updateInvestment(i: Investment) = investmentDao.update(i)
    suspend fun deleteInvestment(i: Investment) = investmentDao.delete(i)

    /** Compute live balance for one account from confirmed transactions. */
    fun computeBalance(account: Account, txns: List<Txn>): Double {
        var bal = account.openingBalance
        for (t in txns) {
            if (t.status != TxnStatus.CONFIRMED) continue
            when (t.type) {
                TxnType.INCOME -> if (t.accountId == account.id) bal += t.amount
                TxnType.EXPENSE -> if (t.accountId == account.id) bal -= t.amount
                TxnType.TRANSFER -> {
                    if (t.accountId == account.id) bal -= t.amount
                    if (t.toAccountId == account.id) bal += t.amount
                }
            }
        }
        return bal
    }

    companion object {
        // Special category names used by automated flows.
        const val CAT_REPAYMENT = "Refund / Repayment"
        const val CAT_SETTLEMENT = "Debt Settlement"
        const val CAT_CARD_PAYMENT = "Credit Card Payment"
        const val CAT_CARD_POINTS = "Card Points Redeemed"
        const val CAT_INVESTMENT = "Investments"
        const val CAT_INVEST_RETURN = "Investment Returns"
        const val CAT_LENT = "Money Lent"
        const val CAT_BORROWED = "Money Borrowed"

        val defaultCategories: List<Category> = listOf(
            // Expenses
            Category(name = "Food", type = TxnType.EXPENSE, icon = "restaurant", colorHex = "#E17055", isDefault = true, sortOrder = 0),
            Category(name = "Groceries", type = TxnType.EXPENSE, icon = "shopping_cart", colorHex = "#00CEC9", isDefault = true, sortOrder = 1),
            Category(name = "Clothing", type = TxnType.EXPENSE, icon = "checkroom", colorHex = "#0984E3", isDefault = true, sortOrder = 2),
            Category(name = "Family", type = TxnType.EXPENSE, icon = "family_restroom", colorHex = "#00B894", isDefault = true, sortOrder = 3),
            Category(name = "Friends / Outing", type = TxnType.EXPENSE, icon = "groups", colorHex = "#6C5CE7", isDefault = true, sortOrder = 4),
            Category(name = "Transport", type = TxnType.EXPENSE, icon = "directions_car", colorHex = "#FDCB6E", isDefault = true, sortOrder = 5),
            Category(name = "Fuel", type = TxnType.EXPENSE, icon = "local_gas", colorHex = "#D63031", isDefault = true, sortOrder = 6),
            Category(name = "Travel", type = TxnType.EXPENSE, icon = "flight", colorHex = "#0984E3", isDefault = true, sortOrder = 7),
            Category(name = "Bills & Utilities", type = TxnType.EXPENSE, icon = "receipt_long", colorHex = "#D63031", isDefault = true, sortOrder = 8),
            Category(name = "Rent", type = TxnType.EXPENSE, icon = "home", colorHex = "#E17055", isDefault = true, sortOrder = 9),
            Category(name = "Shopping", type = TxnType.EXPENSE, icon = "shopping_bag", colorHex = "#E84393", isDefault = true, sortOrder = 10),
            Category(name = "Health", type = TxnType.EXPENSE, icon = "local_hospital", colorHex = "#FF7675", isDefault = true, sortOrder = 11),
            Category(name = "Entertainment", type = TxnType.EXPENSE, icon = "movie", colorHex = "#A29BFE", isDefault = true, sortOrder = 12),
            Category(name = "Subscriptions", type = TxnType.EXPENSE, icon = "subscriptions", colorHex = "#6C5CE7", isDefault = true, sortOrder = 13),
            Category(name = "Education", type = TxnType.EXPENSE, icon = "school", colorHex = "#0984E3", isDefault = true, sortOrder = 14),
            Category(name = "Personal Care", type = TxnType.EXPENSE, icon = "spa", colorHex = "#E84393", isDefault = true, sortOrder = 15),
            Category(name = "Gifts & Donations", type = TxnType.EXPENSE, icon = "card_giftcard", colorHex = "#FAB1A0", isDefault = true, sortOrder = 16),
            Category(name = "EMI & Loans", type = TxnType.EXPENSE, icon = "savings", colorHex = "#636E72", isDefault = true, sortOrder = 17),
            Category(name = "Insurance", type = TxnType.EXPENSE, icon = "shield", colorHex = "#00B894", isDefault = true, sortOrder = 18),
            Category(name = "Pets", type = TxnType.EXPENSE, icon = "pets", colorHex = "#FDCB6E", isDefault = true, sortOrder = 19),
            Category(name = "Sports & Fitness", type = TxnType.EXPENSE, icon = "sports", colorHex = "#00CEC9", isDefault = true, sortOrder = 20),
            Category(name = CAT_INVESTMENT, type = TxnType.EXPENSE, icon = "trending_up", colorHex = "#1AAD7E", isDefault = true, sortOrder = 21),
            Category(name = CAT_CARD_PAYMENT, type = TxnType.EXPENSE, icon = "credit_card", colorHex = "#E17055", isDefault = true, sortOrder = 22),
            Category(name = CAT_SETTLEMENT, type = TxnType.EXPENSE, icon = "handshake", colorHex = "#A29BFE", isDefault = true, sortOrder = 23),
            Category(name = CAT_LENT, type = TxnType.EXPENSE, icon = "handshake", colorHex = "#0984E3", isDefault = true, sortOrder = 24),
            Category(name = "Other", type = TxnType.EXPENSE, icon = "category", colorHex = "#636E72", isDefault = true, sortOrder = 25),
            // Income
            Category(name = "Salary", type = TxnType.INCOME, icon = "payments", colorHex = "#00B894", isDefault = true, sortOrder = 0),
            Category(name = "Business", type = TxnType.INCOME, icon = "store", colorHex = "#0984E3", isDefault = true, sortOrder = 1),
            Category(name = "Freelance", type = TxnType.INCOME, icon = "work", colorHex = "#6C5CE7", isDefault = true, sortOrder = 2),
            Category(name = CAT_REPAYMENT, type = TxnType.INCOME, icon = "undo", colorHex = "#0984E3", isDefault = true, sortOrder = 3),
            Category(name = "Cashback / Rewards", type = TxnType.INCOME, icon = "card_giftcard", colorHex = "#6C5CE7", isDefault = true, sortOrder = 4),
            Category(name = CAT_CARD_POINTS, type = TxnType.INCOME, icon = "star", colorHex = "#FDCB6E", isDefault = true, sortOrder = 5),
            Category(name = CAT_INVEST_RETURN, type = TxnType.INCOME, icon = "trending_up", colorHex = "#1AAD7E", isDefault = true, sortOrder = 6),
            Category(name = CAT_BORROWED, type = TxnType.INCOME, icon = "handshake", colorHex = "#FDCB6E", isDefault = true, sortOrder = 7),
            Category(name = "Interest", type = TxnType.INCOME, icon = "savings", colorHex = "#00CEC9", isDefault = true, sortOrder = 8),
            Category(name = "Other Income", type = TxnType.INCOME, icon = "category", colorHex = "#636E72", isDefault = true, sortOrder = 9)
        )

        suspend fun seedIfEmpty(db: AppDatabase) {
            val repo = Repository(db)
            if (repo.accountDao.count() == 0) {
                val accs = listOf(
                    Account(name = "Bank 1", type = AccountType.BANK, colorHex = "#0984E3", icon = "account_balance", sortOrder = 0),
                    Account(name = "Bank 2", type = AccountType.BANK, colorHex = "#00B894", icon = "account_balance", sortOrder = 1),
                    Account(name = "Bank 3", type = AccountType.BANK, colorHex = "#6C5CE7", icon = "account_balance", sortOrder = 2),
                    Account(name = "Cash", type = AccountType.CASH, colorHex = "#FDCB6E", icon = "payments", sortOrder = 3),
                    Account(name = "SBI PhonePe Credit Card", type = AccountType.CREDIT_CARD, colorHex = "#E17055", icon = "credit_card", includeInTotal = false, sortOrder = 4)
                )
                accs.forEach { repo.accountDao.insert(it) }
            }
            // Add any default categories that don't already exist (covers fresh installs and updates).
            val existing = repo.categoryDao.allOnce().map { it.name }.toHashSet()
            defaultCategories.filter { it.name !in existing }.forEach { repo.categoryDao.insert(it) }
        }
    }
}

File: farhans-wallet-flutter/lib/data/repository.dart
import 'package:sqflite/sqflite.dart';

import 'app_database.dart';
import 'models.dart';

class Repository {
  final AppDatabase _appDb;
  Repository(this._appDb);

  Future<Database> get _db async => _appDb.database;

  // ───────────────────────── Reads ─────────────────────────
  Future<List<Account>> accounts({bool includeArchived = false}) async {
    final db = await _db;
    final rows = await db.query(
      'accounts',
      where: includeArchived ? null : 'archived = 0',
      orderBy: 'sortOrder, id',
    );
    return rows.map(Account.fromMap).toList();
  }

  Future<List<Category>> categories() async {
    final db = await _db;
    final rows = await db.query('categories', orderBy: 'sortOrder, name');
    return rows.map(Category.fromMap).toList();
  }

  Future<List<Txn>> transactions() async {
    final db = await _db;
    final rows = await db.query('transactions', orderBy: 'dateTime DESC');
    return rows.map(Txn.fromMap).toList();
  }

  Future<List<Debt>> debts() async {
    final db = await _db;
    final rows = await db.query('debts', orderBy: 'settled, createdAt DESC');
    return rows.map(Debt.fromMap).toList();
  }

  Future<List<Investment>> investments() async {
    final db = await _db;
    final rows = await db.query('investments', orderBy: 'sold, createdAt DESC');
    return rows.map(Investment.fromMap).toList();
  }

  // ───────────────────────── Writes ─────────────────────────
  Future<int> upsertAccount(Account a) async {
    final db = await _db;
    if (a.id == null) return db.insert('accounts', a.toMap()..remove('id'));
    await db.update('accounts', a.toMap(), where: 'id = ?', whereArgs: [a.id]);
    return a.id!;
  }

  Future<void> deleteAccount(int id) async {
    final db = await _db;
    await db.delete('accounts', where: 'id = ?', whereArgs: [id]);
  }

  Future<int> upsertCategory(Category c) async {
    final db = await _db;
    if (c.id == null) return db.insert('categories', c.toMap()..remove('id'));
    await db.update('categories', c.toMap(), where: 'id = ?', whereArgs: [c.id]);
    return c.id!;
  }

  Future<void> deleteCategory(int id) async {
    final db = await _db;
    await db.delete('categories', where: 'id = ?', whereArgs: [id]);
  }

  Future<int> upsertTxn(Txn t) async {
    final db = await _db;
    if (t.id == null) return db.insert('transactions', t.toMap()..remove('id'));
    await db.update('transactions', t.toMap(), where: 'id = ?', whereArgs: [t.id]);
    return t.id!;
  }

  Future<void> deleteTxn(int id) async {
    final db = await _db;
    await db.delete('transactions', where: 'id = ?', whereArgs: [id]);
  }

  Future<int> upsertDebt(Debt d) async {
    final db = await _db;
    if (d.id == null) return db.insert('debts', d.toMap()..remove('id'));
    await db.update('debts', d.toMap(), where: 'id = ?', whereArgs: [d.id]);
    return d.id!;
  }

  Future<void> deleteDebt(int id) async {
    final db = await _db;
    await db.delete('debts', where: 'id = ?', whereArgs: [id]);
  }

  Future<int> upsertInvestment(Investment i) async {
    final db = await _db;
    if (i.id == null) return db.insert('investments', i.toMap()..remove('id'));
    await db.update('investments', i.toMap(), where: 'id = ?', whereArgs: [i.id]);
    return i.id!;
  }

  Future<void> deleteInvestment(int id) async {
    final db = await _db;
    await db.delete('investments', where: 'id = ?', whereArgs: [id]);
  }

  Future<int?> categoryIdByName(String name, TxnType type) async {
    final db = await _db;
    final rows = await db.query('categories',
        columns: ['id'], where: 'name = ? AND type = ?', whereArgs: [name, type.name], limit: 1);
    if (rows.isEmpty) return null;
    return rows.first['id'] as int;
  }

  Future<bool> upiRefExists(String ref) async {
    final db = await _db;
    final rows = await db.query('transactions',
        columns: ['id'], where: 'upiRef = ?', whereArgs: [ref], limit: 1);
    return rows.isNotEmpty;
  }

  // ───────────────────────── Backup / restore ─────────────────────────
  Future<Map<String, dynamic>> exportAll() async {
    final db = await _db;
    return {
      'app': 'pocket_flow',
      'version': 1,
      'exportedAt': DateTime.now().millisecondsSinceEpoch,
      'accounts': await db.query('accounts'),
      'categories': await db.query('categories'),
      'transactions': await db.query('transactions'),
      'debts': await db.query('debts'),
      'investments': await db.query('investments'),
    };
  }

  /// Replaces all current data with the contents of a backup file.
  Future<void> importAll(Map<String, dynamic> data) async {
    final db = await _db;
    await db.transaction((txn) async {
      for (final t in ['transactions', 'debts', 'investments', 'categories', 'accounts']) {
        await txn.delete(t);
      }
      for (final table in ['accounts', 'categories', 'transactions', 'debts', 'investments']) {
        final rows = (data[table] as List?) ?? const [];
        for (final row in rows) {
          await txn.insert(table, Map<String, Object?>.from(row as Map),
              conflictAlgorithm: ConflictAlgorithm.replace);
        }
      }
    });
  }

  /// Live balance for one account. For credit cards the balance represents the
  /// amount OWED (outstanding): spending/lending raises it, payments reduce it.
  static double computeBalance(Account account, List<Txn> txns) {
    final card = account.type == AccountType.creditCard;
    var bal = account.openingBalance;
    for (final t in txns) {
      if (t.status != TxnStatus.confirmed) continue;
      switch (t.type) {
        case TxnType.income:
          if (t.accountId == account.id) bal += card ? -t.amount : t.amount;
          break;
        case TxnType.expense:
          if (t.accountId == account.id) bal += card ? t.amount : -t.amount;
          break;
        case TxnType.transfer:
          if (t.accountId == account.id) bal += card ? t.amount : -t.amount;
          if (t.toAccountId == account.id) bal += card ? -t.amount : t.amount;
          break;
      }
    }
    return bal;
  }
}
